### PR TITLE
Part 1. Hook interfaces

### PR DESCRIPTION
**Pull request by 4lexvav**
Added basic hook interfaces. In the following tasks, we can make further changes to the interfaces.
Next part: #7

### CODE REVIEW

File: hooks/invocation/invocation.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package invocation
+
+import "time"
+
+type Action int
+
+const (
+	Nop Action = iota
+	Update
+	Reject
```
Comment (SyntaxNode) [inline hooks/invocation/invocation.go]: I'm curious what you plan to use the Action code for? We will know if no action is taken if there are no mutations and from today's meeting we expect to pass back a NBR (no bid response) code is it's rejected.

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/prebid/prebid-server
 
-go 1.16
+go 1.19
```
Comment (SyntaxNode) [inline go.mod]: Please submit this change as a separate PR to master. There are other files which need to be updated too, such as the github test workflows and the docker files.

File: hooks/module.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package hooks
+
+type Module interface {
+	Name() string
```
Comment (SyntaxNode) [inline hooks/module.go]: Does the Name need to be a part of the interface? .. or can we link it from the Builders map?

File: hooks/module.go [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+	Hooks() []interface{}
```
Comment (SyntaxNode) [inline hooks/module.go]: The Hooks() function is not necessarily if we represent each hook as an interface. This was my intention from our discussions. For example, the bid request hook interface can be defined as: something like

```
type BidRequestHook interface {
  HandleBidRequest(...)
}
```

The module passively supports the hook if it implements that interface.

File: hooks/stages/entrypoint_test.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+type mockNopEntrypointHook struct{}
+
+func (e mockNopEntrypointHook) Code() string {
+	return "MockedNopEntrypointHook"
+}
+
+func (e mockNopEntrypointHook) Call(ctx context.Context, invocationContext invocation.Context, payload ImmutableEntrypointPayload) (invocation.HookResult[MutableEntrypointPayload], error) {
+	_ = jsonparser.Delete(payload.Body, "name")
+	_, _ = jsonparser.Set(payload.Body, []byte(`"bar"`), "foo")
```
Comment (SyntaxNode) [inline hooks/stages/entrypoint_test.go]: Please avoid using jsonparser Delete and Set methods. They are indicated as not production ready by the library author. We've tried to use them in production before and encountered situations where they panic.

They should be ok for tests, but might give the wrong impression that we as the PBS project support the use. We strictly use jsonparser for read only operations.

File: hooks/stages/entrypoint_test.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+	return invocation.HookResult[MutableEntrypointPayload]{Action: invocation.Nop}, nil
+}
+
+func TestExecuteEntrypointHook_CanApplyHookMutations(t *testing.T) {
+	entrypointHooks = []EntrypointHook{mockUpdateEntrypointHook{}, mockUpdateEntrypointHook{}}
+
+	ctx := context.TODO()
```
Comment (SyntaxNode) [inline hooks/stages/entrypoint_test.go]: Nitpick: For tests we should use `context.Background()` instead of `context.TODO()`. TODO is not intended for use in tests.

File: hooks/stages/allbidresponse.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package stages
+
+import (
+	"context"
+
+	"github.com/prebid/prebid-server/hooks/invocation"
+)
+
+type AllBidResponsesHook interface {
+	Code() string
```
Comment (SyntaxNode) [inline hooks/stages/allbidresponse.go]: What is the purpose of `.Code()`?

**Review by SyntaxNode — COMMENTED** [review]
I'm excited to see this take shape.

File: hooks/module.go (hunk 3 above)
Comment (4lexvav) [reply hooks/module.go]: Yes, we can use Builders map instead.

File: hooks/module.go (hunk 4 above)
Comment (4lexvav) [reply hooks/module.go]: I think I got your idea, in this case for now we will get rid of the Module interface at all.

File: hooks/invocation/invocation.go (hunk 1 above)
Comment (4lexvav) [reply hooks/invocation/invocation.go]: We tried to implement this as similar as possible to the PBS-Java code, but I agree with you, we do not need the Update/Nop actions, this can be concluded from the presence of mutations.
Rejection itself we will implement in the following tasks.

File: hooks/stages/allbidresponse.go (hunk 7 above)
Comment (4lexvav) [reply hooks/stages/allbidresponse.go]: We assume the module could provide several implementations of the same hook interface. So admin when configuring the hook execution plan can specify the code of the hook he wants to use.
So we added hook code to distinguish different hooks of the same hook interface (similar to PBS-Java).
But, you are right, we can remove the `Code` method from the Hook interface, we can use map for this. You will have more context after we add PRs for module loading and hook invocation tasks.

**Comment by 4lexvav** [comment]
✅ merged
